core/desugarLAttributes.ml

```diff
@@ -78,33 +86,36 @@ let desugar_lonevent : phrasenode -> phrasenode =
                          rhs]
     | _ -> assert false
   in function
-    | Xml (tag, attrs, attrexp, children)
+    | { node=Xml (tag, attrs, attrexp, children); pos }
         when exists (fst ->- start_of ~is:"l:on") attrs ->
         let lons, others = partition (fst ->- start_of ~is:"l:on") attrs in
         let idattr =
           ("key",
            [apply "registerEventHandlers"
                   [list (List.map (event_handler_pair) lons)]]) in
-          Xml (tag, idattr::others, attrexp, children)
+          WithPos.make ~pos (Xml (tag, idattr::others, attrexp, children))
     | e -> e
 
-let desugar_lnames (p : phrasenode) : phrasenode * (string * string) StringMap.t =
+let desugar_lnames (p : phrase) : phrase * (string * string) StringMap.t =
   let lnames = ref StringMap.empty in
   let add lname (id,name) = lnames := StringMap.add lname (id,name) !lnames in
-  let attr : string * phrase list -> (string * phrase list) list = function
-    | "l:name", [{node=Constant (Constant.String v); _}] ->
-        let id, name = fresh_names () in
-          add v (id,name);
-          [("name", [constant_str name]);
-           ("id"  , [constant_str id  ])]
-    | "l:name", _ -> failwith ("Invalid l:name binding")
-    | a -> [a] in
-  let rec aux : phrasenode -> phrasenode  = function
-    | Xml (tag, attrs, attrexp, children) ->
-        let attrs = concat_map attr attrs in
+  let attr : Position.t -> string * phrase list -> (string * phrase list) list =
+    fun pos attribute ->
+    match attribute with
+      | "l:name", [{node=Constant (Constant.String v); _}] ->
+          let id, name = fresh_names () in
+            add v (id,name);
+            [("name", [constant_str name]);
+             ("id"  , [constant_str id  ])]
+      | "l:name", _ ->
+          desugaring_error pos "The value l:name binding must be a string constant"
+      | a -> [a] in
+  let rec aux : phrase -> phrase  = function
+    | { node=Xml (tag, attrs, attrexp, children); pos } ->
+        let attrs = concat_map (attr pos) attrs in
         let children =
-          List.map (fun {node;pos} -> with_pos pos (aux node)) children in
-          Xml (tag, attrs, attrexp, children)
+          List.map (fun child -> aux child) children in
```

Comment:
Would OCaml allow `List.map aux children`?

Reply:
Yes, why I didn't eta-contract that remains a mystery to me.